WebCodecs AudioEncoder implementation

philn opened the pull request:
#### 08a37f0893cdc5a25ccc5eadb855f7751949635a
<pre>
WebCodecs AudioEncoder implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=261107">https://bugs.webkit.org/show_bug.cgi?id=261107</a>

Reviewed by Youenn Fablet.

This patch exposes AudioEncoder and its implementation for ports enabling the WebCodecsAudioEnabled
setting.

Canonical link: <a href="https://commits.webkit.org/267963@main">https://commits.webkit.org/267963@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d39393d36e3818485e25cd4afdea2b09025de757

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18500 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19070 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17007 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18365 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21797 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18658 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18968 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18388 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18619 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15818 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20886 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15843 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16571 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23081 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16862 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20964 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14678 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16401 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4335 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17160 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -0,0 +1,437 @@
+
+        if (flacConfig->compressLevel > 8)
+            return false;
+    }
```
I think it would be a good idea to add an `isValid()` method to the structures instead of having this check here.

calvaris commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -0,0 +1,437 @@
+
+    if (m_hasNewActiveConfiguration) {
+        m_hasNewActiveConfiguration = false;
+        // FIXME: Provide more accurate decoder configuration
```
.

calvaris commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -0,0 +1,437 @@
+        return;
+    }
+
+    m_pendingFlushPromises.append(promise.copyRef());
```
It seems you don't use the promise after the append. You could just move it in instead of copying the ref since you own it.

calvaris commented on Source/WebCore/platform/AudioEncoder.cpp:
```diff
@@ -0,0 +1,78 @@
+#if USE(GSTREAMER)
+    auto& scanner = GStreamerRegistryScanner::singleton();
+    result = scanner.isCodecSupported(GStreamerRegistryScanner::Configuration::Encoding, codec.toString());
+#endif
```
This should be moved to the platform class.

philn replied:
we're already in platform/ and at this point there is no platform-specific encoder created yet.

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -0,0 +1,364 @@
+    g_signal_connect(pad.get(), "notify::caps", G_CALLBACK(+[](GObject* pad, GParamSpec*, gpointer userData) {
+        auto* encoder = static_cast<WeakPtr<GStreamerInternalAudioEncoder>*>(userData);
+        if (!encoder)
+            return;
```
This check is going to return always true as you're creating a newed WeakPtr object below and passing it as userData. auto* will be as stated in the static_cast, so you need to check of what is inside the WeakPtr is valid.

Maybe something like:

`WeakPtr<GStreamerInternalAudioEncoder> encoder(*static_cast<WeakPtr<GStreamerInternalAudioEncoder>*>(userData));`

You'd also be able to move the weakPtr into the lambda closure.

And very important, contrary to what I said in another review, you need to use g_signal_connect_data to be able to provide the proper GDestroyNotify function that would just call delete on it. Before, I missed that the signal could be called more than once.

youennf commented on Source/WebCore/Modules/webcodecs/AacEncoderConfig.h:
```diff
@@ -0,0 +1,38 @@
+
+enum class AacBitstreamFormat {
+    Aac,
+    Adts
```
Bool or uint8_t

youennf commented on Source/WebCore/Modules/webcodecs/AacEncoderConfig.h:
```diff
@@ -0,0 +1,38 @@
+};
+
+struct AacEncoderConfig {
+    using BitstreamFormat = AacBitstreamFormat;
```
Needed?

philn replied:
As per EWS results, yes :)

youennf commented on Source/WebCore/Modules/webcodecs/FlacEncoderConfig.h:
```diff
@@ -0,0 +1,44 @@
+namespace WebCore {
+
+struct FlacEncoderConfig {
+    bool isValid()
```
Can we add the link to the WebCodecs Flac registry corresponding section?

youennf commented on Source/WebCore/Modules/webcodecs/OpusEncoderConfig.h:
```diff
@@ -0,0 +1,64 @@
+};
+
+struct OpusEncoderConfig {
+    bool isValid()
```
Can we add the link to the WebCodecs opus registry  corresponding section?

youennf commented on Source/WebCore/Modules/webcodecs/OpusEncoderConfig.h:
```diff
@@ -0,0 +1,64 @@
+        return true;
+    }
+
+    using BitstreamFormat = OpusBitstreamFormat;
```
Needed?

philn replied:
Ditto

youennf commented on Source/WebCore/Modules/webcodecs/OpusEncoderConfig.h:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    using BitstreamFormat = OpusBitstreamFormat;
+    OpusBitstreamFormat format;
```
Default value?

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoder.cpp:
```diff
@@ -0,0 +1,425 @@
+    if (StringView(config.codec).trim(isASCIIWhitespace<UChar>).isEmpty())
+        return false;
+
+    if (auto opusConfig = config.opus) {
```
These checks will probably need to move so that they trigger not supported in the future.

philn replied:
Should I add a FIXME about this?

youennf replied:
Sounds good

youennf commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -0,0 +1,375 @@
+{
+    gstEncoderWorkQueue().dispatch([frame = WTFMove(frame), encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
+        auto result = encoder->encode(WTFMove(frame));
+        if (encoder->isClosed())
```
Should callback be called here?

philn replied:
Hmm, I don't think so.

youennf replied:
OK, there are cases where we do not always call the callback.
This seems fine.

youennf replied:
Yeah, we are not very consistent here, I guess it is fine.

youennf commented on Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp:
```diff
@@ -0,0 +1,375 @@
+
+void GStreamerAudioEncoder::encode(RawFrame&& frame, EncodeCallback&& callback)
+{
+    gstEncoderWorkQueue().dispatch([frame = WTFMove(frame), encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
```
I would think callback should be called on the right thread, executing it in the encoder worker queue might not be good.

philn replied:
Hmm, I'm not sure which thread you're talking about :)

youennf replied:
I think we usually call postTask callback to hop to the right thread and inside call the callback.

philn replied:
I'll give it a try... If that works the same change will be needed in the gst video encoder I think. Thanks for the suggestion!

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsAudioEncoderConfig.h:
```diff
@@ -0,0 +1,56 @@
+struct WebCodecsAudioEncoderConfig {
+    String codec;
+    size_t sampleRate;
+    size_t numberOfChannels;
```
Why aren't these two optional?

philn replied:
Good point, they're not `required` attributes in the spec.

philn committed:
WebCodecs AudioEncoder implementation
https://bugs.webkit.org/show_bug.cgi?id=261107

Reviewed by Youenn Fablet.

This patch exposes AudioEncoder and its implementation for ports enabling the WebCodecsAudioEnabled
setting.

Canonical link: https://commits.webkit.org/267963@main
